### ISSUE TITLE
Bug when using MSBuild with custom target since NET5.0 update

### ISSUE THREAD

**Issue by theolivenbaum** [dotnet/msbuild#5875]
### Issue Description
I'm hitting a bug when consuming a custom project SDK target (http://nuget.org/packages/h5.target):

`````
Invalid static method invocation syntax: "[MSBuild]::GetTargetFrameworkVersion('$(TargetFramework)', 2)". Method '[MSBuild]::GetTargetFrameworkVersion' not found. Static method invocation should be of the form: $([FullTypeName]::Method()), e.g. $([System.IO.Path]::Combine(`a`, `b`)). Check that all parameters are defined, are of the correct type, and are specified in the right order.  C:\Program Files\dotnet\sdk\5.0.100\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.TargetFrameworkInference.targets
`````

The code in the targets file that seems to be triggering this issue is the following:

`````xml
  <PropertyGroup Condition="'$(TargetFramework)' != '' and ('$(TargetFrameworkIdentifier)' == '' or '$(TargetFrameworkVersion)' == '')">

    <TargetFrameworkIdentifier>$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)'))</TargetFrameworkIdentifier>
    <TargetFrameworkVersion>v$([MSBuild]::GetTargetFrameworkVersion('$(TargetFramework)', 2))</TargetFrameworkVersion>
  </PropertyGroup>
`````

If I manually force the properties to the expected (or dummy) values, so that we avoid the GetTargetFrameworkIdentifier, GetTargetFrameworkVersion, GetTargetPlatformIdentifier and GetTargetPlatformVersion calls above:

`````xml
    <TargetFrameworkIdentifier>.NETStandard</TargetFrameworkIdentifier>
    <TargetFrameworkVersion>v2.0</TargetFrameworkVersion>
    <TargetPlatformIdentifier>h5</TargetPlatformIdentifier>
    <TargetPlatformVersion>2.0</TargetPlatformVersion>
`````

Then the compilation using MSBuild works fine. 

Any ideas?

**Comment by BenVillalobos** [comment]
Team Triage: Can you confirm the versions of MSBuild and SDK you're using? It sounds like you're using an older MSBuild with a newer SDK.

**Comment by theolivenbaum** [comment]
I thought that was the case, but if I enumerate the MSBuild version being
found and selected by the MSBuildLocator, it correctly points to the
5.0.100 one under the dotnet/sdks folder.

**Comment by BenVillalobos** [comment]
Very strange, can you post the output of `dotnet msbuild --version`  from your 5.0.100 sdk?

I can't seem to repro it from sdk version `5.0.100-rc.2.20479.15` that's using MSBuild version `16.8.0-preview-20475-05+aed5e7ed0`

**Comment by hellodeibu** [comment]
I'm getting the same issue for a project that is targeting net45:

```
/usr/share/dotnet/sdk/5.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(5,5): Error MSB4186: Invalid static method invocation syntax: "[MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')". Method '[MSBuild]::GetTargetFrameworkIdentifier' not found. Static method invocation should be of the form: $([FullTypeName]::Method()), e.g. $([System.IO.Path]::Combine(`a`, `b`)). Check that all parameters are defined, are of the correct type, and are specified in the right order. (MSB4186) (SMAPI.ModBuildConfig.Analyzer)
/usr/share/dotnet/sdk/5.0.100/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.TargetFrameworkInference.targets(5,5): Error MSB4186: Invalid static method invocation syntax: "[MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')". Method '[MSBuild]::GetTargetFrameworkIdentifier' not found. Static method invocation should be of the form: $([FullTypeName]::Method()), e.g. $([System.IO.Path]::Combine(`a`, `b`)). Check that all parameters are defined, are of the correct type, and are specified in the right order. (MSB4186) (SMAPI.Toolkit.CoreInterfaces)
```
```bash
> dotnet --version
5.0.100
> msbuild -version
16.6.0.32601%
```

The Project Options view in Monodevelop (7.8.4 (build 2)) actually shows the same oddness:
![image](https://user-images.githubusercontent.com/18850/98918317-cd5c1300-2510-11eb-8b1e-ca7ef43055cc.png)

**Comment by emmanuel128** [comment]
Hi, 

I'm getting the same error as @hellodeibu on Visual Studio for Mac. I have a project is targeting net48 and the Xamarin project is targeting netstandard2.0.

Does anyone have a solution?

**Comment by BenVillalobos** [comment]
@emmanuel128 @hellodeibu could you post specifically your `dotnet msbuild --version`? Info? As typically `msbuild -version` points to the msbuild installed by visual studio rather than the msbuild packed along with the sdk

**Comment by hellodeibu** [comment]
@BenVillalobos apologies for not including that too:

```bash
> dotnet --version
5.0.100
> msbuild -version
16.6.0.32601%
> dotnet msbuild -version
16.8.0.52002% 
```
Just in case this helps too: this is on Ubuntu 20.04 and with MonoDevelop 7.8.4 (build 2) installed.

**Comment by theolivenbaum** [comment]
Same here, using the latest version from the VS2019 installation.

`````bash
**********************************************************************
** Visual Studio 2019 Developer Command Prompt v16.8.0
** Copyright (c) 2020 Microsoft Corporation
**********************************************************************

>msbuild -version
Microsoft (R) Build Engine version 16.8.1+bd2ea1e3c for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

16.8.1.52902
>dotnet msbuild -version
Microsoft (R) Build Engine version 16.8.0+126527ff1 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

16.8.0.52002
>dotnet --version
5.0.100
`````

Also in my case this happens when calling MSBuild programmatically with the [h5 compiler](h5.rocks).

**Comment by BenVillalobos** [comment]
We're curious if this may be a GAC issue. Please try repairing Visual Studio. This will at least eliminate that possibility.

**Comment by BenVillalobos** [comment]
For those on Mono, this might be because Mono needs to be updated. Here is an issue tracking this: https://github.com/mono/mono/issues/20250

**Comment by theolivenbaum** [comment]
Not Mono on my case as running on Windows. Already tried the VS repair but didn't fix it either.

And same issue happens when running dotnet build on linux so don't think it's the case.

**Comment by BenVillalobos** [comment]
Tried to repro again and not seeing any issues:

`dotnet msbuild -version`
```
Microsoft (R) Build Engine version 16.8.0+126527ff1 for .NET       
Copyright (C) Microsoft Corporation. All rights reserved.

16.8.0.52002
```
`dotnet --version`
```
5.0.100
```

@theolivenbaum could you try building a tiny project (via `dotnet build` on your 5.0.100 sdk) that uses `[MSBuild]::GetTargetFrameworkIdentifier`? This should eliminate whether it could be MSBuildLocator in your scenario.

The project I'm building to try and repro this is simply:
```xml
<Project>
<PropertyGroup>
    <TargetFrameworkIdentifier>.NETStandard</TargetFrameworkIdentifier>
    <TargetFrameworkVersion>v2.0</TargetFrameworkVersion>
    <TargetPlatformIdentifier>h5</TargetPlatformIdentifier>
    <TargetPlatformVersion>2.0</TargetPlatformVersion>
</PropertyGroup>
  <PropertyGroup>
    <TargetFrameworkIdentifier>$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)'))</TargetFrameworkIdentifier>
    <TargetFrameworkVersion>v$([MSBuild]::GetTargetFrameworkVersion('$(TargetFramework)', 2))</TargetFrameworkVersion>
  </PropertyGroup>

  <Target Name="HelloWorld">
    <Message Text="Hello World!" Importance="High"/>
  </Target>
</Project>
```

Note that this _does_ fail with your error if I `global.json` myself to do `3.1.404`, which is expected as that version of MSBuild would not have the functions being called.

You mentioned MSBuildLocator is finding the correct MSBuild, can you verify this? I ask because this issue (should) only come up if you were using an older MSBuild wth a newer sdk somehow.

What may be the most helpful in this scenario:
1. Providing a minimal repro of your specific scenario.
2. Providing binlogs of your build.

**Comment by theolivenbaum** [comment]
Hi @BenVillalobos 

Tried to reproduce the issue today - but I'm not seeing it anymore, even when reverting the recent commits on the h5 compiler and SDK targets. Wondering if the update to VS 2019 16.8.1 that came immediately after 16.8.0 fixed the problem meanwhile and this was indeed a GAC cache issue...

**Comment by BenVillalobos** [comment]
@theolivenbaum It happens 🙂

Closing this issue out. For those running on mono, see this issue: https://github.com/mono/mono/issues/20250

**Comment by granadacoder** [comment]
For future readers.

(On a Mac).


I updated (added) (only) dotnetcore SDK 5.0, and started getting this error.

`error msb4186: invalid static method invocation syntax: "[msbuild]::gettargetframeworkidentifier('$(targetframework)')". method '[msbuild]::gettargetframeworkidentifier' not found. static method invocation should be of the form: $([fulltypename]::method()), e.g. $([system.io.path]::combine(`a`, `b`)). check that all parameters are defined, are of the correct type, and are specified in the right order. (msb4186)
`

Our work machines are fairly "locked down", so I do this upgrades one by one to easy trouble shooting if anything goes wrong.

Below image was what was shown when I did a "upgrade" (aka, below is the newer version at the time of writing)
Unfortunately, I did not capture the original version.

![image](https://user-images.githubusercontent.com/5487824/110473223-4da01880-80ac-11eb-96c4-33884b2b20d6.png)


After I did a Visual Studio Update (March 2021 timeline), I opened up my solution, and everything was ok in the nuget restore world.
